src/StreamChannel.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace Amp\Sync;
+
+use Amp\ByteStream\ReadableStream;
+use Amp\ByteStream\StreamException;
+use Amp\ByteStream\WritableStream;
+use Amp\Cancellation;
+use Amp\Pipeline\ConcurrentIterator;
+use Amp\Pipeline\Pipeline;
+use Amp\Serialization\Serializer;
+use Amp\Sync\Internal\ChannelParser;
+
+/**
+ * An asynchronous channel for sending data between threads and processes.
+ *
+ * Supports full duplex read and write.
+ *
+ * @template TReceive
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class StreamChannel implements Channel
+{
+    private ReadableStream $read;
+
+    private WritableStream $write;
+
+    private ChannelParser $parser;
+
+    /** @var ConcurrentIterator<TReceive> */
+    private ConcurrentIterator $iterator;
+
+    /**
+     * Creates a new channel from the given stream objects. Note that $read and $write can be the same object.
+     *
+     * @param ReadableStream $read
+     * @param WritableStream $write
+     * @param Serializer|null $serializer
+     */
+    public function __construct(ReadableStream $read, WritableStream $write, ?Serializer $serializer = null)
+    {
+        $this->read = $read;
+        $this->write = $write;
+
+        $received = new \SplQueue();
+        $this->parser = $parser = new ChannelParser(\Closure::fromCallable([$received, 'push']), $serializer);
+
+        $this->iterator = Pipeline::fromIterable(static function () use ($read, $received, $parser): \Generator {
+            while (true) {
+                try {
+                    $chunk = $read->read();
+                } catch (StreamException $exception) {
+                    throw new ChannelException(
+                        "Reading from the channel failed. Did the context die?",
+                        0,
+                        $exception,
+                    );
+                }
+
+                if ($chunk === null) {
+                    return;
+                }
+
+                $parser->push($chunk);
+
+                while (!$received->isEmpty()) {
+                    yield $received->shift();
+                }
```

Comment:
Should we use `Queue` here, so we can allow a configurable buffer size?

Reply:
Let's leave it for now and revisit this some other time.